runtime/lua/vim/lsp/util.lua

```diff
@@ -99,13 +99,21 @@ function M.apply_text_edits(text_edits, bufnr)
   local start_line, finish_line = math.huge, -1
   local cleaned = {}
   for i, e in ipairs(text_edits) do
+    local start_row = e.range.start.line
+    local start_col = e.range.start.character
+    local start_bline = api.nvim_buf_get_lines(0, start_row, start_row+1, true)[1]
```

Comment:
I think this will fail if the buffer is not loaded which could be the  case for apply_text_edits

Reply:
Right, I (now) remember that issue from the fix for vim-lsp. 

That's the final straw to push me to refactor -- what would be the lua equivalent of `readfile`?

Reply:
Actually, that would be a deeper problem with the current code -- `apply_text_edits` should take a `bufname`, not a `bufnr`.

Reply:
On the other hand, the upstream function calls
```lua
local function uri_to_bufnr(uri)
  return vim.fn.bufadd((uri_to_fname(uri)))
end
```
Doesn't that load the buffer if it isn't already opened?

Reply:
Still shouldn't be `0` instead of `bufnr`...

Reply:
OK, I still get an error if the edit is applied to a different file than the currently opened one, but the buffer is in fact loaded after the rename, it's just hidden (i.e., `buf bufnr` shows the correct file, but `nvim_buf_is_loaded(bufnr)` returns `false`). @justink is there a way to get lines from a hidden buffer?

Reply:
Never mind, there was already code to that effect in the function, I just had to move it up 🤦